Fixed #29973 -- Added compilemessages --ignore option.

rsiemens opened the pull request:
https://code.djangoproject.com/ticket/29973

charettes commented on django/core/management/commands/compilemessages.py:
```diff
@@ -46,10 +46,16 @@ def add_arguments(self, parser):
     def handle(self, **options):
         locale = options['locale']
         exclude = options['exclude']
+        ignore = options['ignore']
```
Convert to a `set` since it will only be used for containment checks.

charettes commented on django/core/management/commands/compilemessages.py:
```diff
@@ -65,6 +71,7 @@ def handle(self, **options):
 
         # Walk entire tree, looking for locale directories
         for dirpath, dirnames, filenames in os.walk('.', topdown=True):
+            dirnames[:] = [d for d in dirnames if d not in ignore]
```
Shouldn't this consider `dirpath` as well?

charettes commented on django/core/management/commands/compilemessages.py:
```diff
@@ -65,6 +71,7 @@ def handle(self, **options):
         for dirpath, dirnames, filenames in os.walk('.', topdown=True):
+            dirnames[:] = [d for d in dirnames if d not in ignore]
             for dirname in dirnames:
                 if dirname == 'locale':
```
You can avoid the double iteration over `dirnames` and list creation by simply adjusting the if below.

```suggestion
                 if dirname == 'locale' and dirname not in ignore:
```

This change made me realize you probably want to consider `dirpath` unless I'm missing something as only `dirname == 'locale'` will be considered anyway? At least that's what the reporter of the ticket seems to be trying to do by passing `--ignore .tox`. You might want to have a look at [how `makemessages --ignore` is implemented](https://github.com/django/django/blob/4bc008028c33cea5307a08c401778fe503e063a4/django/core/management/commands/makemessages.py#L457-L477).

rsiemens replied:
You are right, we can definitely avoid the double iteration. I think we still want to prune the `dirnames` list on a matching `dirname` though otherwise it will recurse into that dir to look for a nested locale directory.
```
./
 |- locale/ 
 |- foo/
  |- locale/
```

We can get the same result checking the path like you suggested, but it would add extra iterations of walk as it would still navigate further down the directory.

rsiemens commented:
Hey @charettes when you have some time can you take a look at it again and provide any thoughts? Thanks!

charettes commented on tests/i18n/test_compilation.py:
```diff
@@ -188,3 +188,49 @@ def test_app_locale_compiled(self):
+        call_command('compilemessages', ignore=['cache'], stdout=StringIO())
+        self.assertTrue(self.all_exist('locale', ['en', 'fr', 'it']))
+        self.assertTrue(self.none_exist(self.CACHE_DIR, ['en', 'fr', 'it']))
+        self.assertTrue(self.all_exist(self.NESTED_DIR, ['en', 'fr', 'it']))
```
Given these methods are all wrapped in `assertTrue` calls they should probably be converted to `assert_all_exists` and `assert_none_exists` that perform the assertion themselves.

e.g.

```python
def assert_all_exists(self, dir, langs):
    self.assertTrue(all(
        os.path.exists(self.MO_FILE % (dir, lang)) for lang in langs
    ))

charettes commented on tests/i18n/test_compilation.py:
```diff
@@ -188,3 +188,49 @@ def test_app_locale_compiled(self):
+        return all(results)
+
+    def test_one_locale_dir_ignored(self):
+        call_command('compilemessages', ignore=['cache'], stdout=StringIO())
```
You can pass `verbosity=0` instead to completely silence the command instead of creating an unused `StringIO` container.

charettes commented on django/core/management/utils.py:
```diff
@@ -122,3 +123,32 @@ def get_command_line_option(argv, option):
+    return norm_patterns
+
+
+def is_ignored(path, ignore_patterns):
```
This would benefit from a more detailed name since it's declared in a quite generic `utils` module.

What about `is_path_ignored`?

charettes commented on django/core/management/utils.py:
```diff
@@ -122,3 +123,32 @@ def get_command_line_option(argv, option):
         return options.value
+
+
+def normalize_patterns(patterns):
```
Ditto about the naming, what about `normalize_path_patterns`?

carltongibson commented on tests/i18n/test_compilation.py:
```diff
@@ -188,3 +188,45 @@ def test_app_locale_compiled(self):
         self.assertTrue(os.path.exists(self.APP_MO_FILE))
+
+
+class IgnoreDirectoryCompilationTest(MessageCompilationTests):
```
Just on this test case: would it work to group the tests using `subTest()`? (They seem quite repetitive.)

rsiemens replied:
While I agree these tests are somewhat repetitive, I ended up opting against using `subTest` and leaving this as is for the following reasons.
- Each test depends on having the directory tree in a clean state which is reset provided by this [addCleanUp](https://github.com/django/django/blob/master/tests/i18n/utils.py#L54) which is called after `tearDown`.
- I think the tests would lose some clarity as the assertions differ slightly, like expecting all to exist [here](https://github.com/django/django/pull/10716/files#diff-30861505e59f9567c5ebc73f6f91e999R221) vs expecting none to exist [here](https://github.com/django/django/pull/10716/files#diff-30861505e59f9567c5ebc73f6f91e999R233).

I'm open to changing if you feel strongly about this though.

rsiemens commented:
Thanks for the review @carltongibson!
I'll aim to have all these items addressed early this week.

rsiemens commented:
@carltongibson, hope you are doing well. I've updated the PR with the following changes:

- Rebase on top of master and adjusted the release notes to for 3.0
- Updated the refactored [is_ignored](https://github.com/django/django/pull/10716/files#diff-e25be54b3794c16fee263cfffae38a06R151) function and [tests](https://github.com/django/django/pull/10716/files#diff-30861505e59f9567c5ebc73f6f91e999R198) to use `pathlib.Path`. I didn't the adjust `makemessages` or `compilemessages` commands to use it though as that seemed like the changes would grow a little out of scope of this PR. If you disagree though I can make those changes here.
- Added some simple unittests for the new helper functions
- Break the PR into separate commits
  - Refactor makemessages to create new helpers [ce77f8c](https://github.com/django/django/pull/10716/commits/ce77f8cb9a70a63fea8a169aa428a01166406ddd)
  - Add new compilemessages option --ignore [57066da](https://github.com/django/django/pull/10716/commits/57066da97d0490edaaa64c70ec3e3310063de08e)
